core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/BootMonitors.scala

```diff
@@ -119,6 +134,25 @@ object BootMonitors extends LazyLogging {
     system.actorOf(BucketDeletionMonitor.props(slickDataSource, gcsDAO, 10 seconds, 6 hours))
   }
 
+  private def startAvroUpsertMonitor(system: ActorSystem, workspaceService: UserInfo => WorkspaceService, googleServicesDAO: GoogleServicesDAO, samDAO: SamDAO, googleStorage: GoogleStorageService[IO], googlePubSubDAO: GooglePubSubDAO, avroUpsertPubSubProject: GoogleProject, avroUpsertPubSubTopic: String, avroUpsertPubSubSubscription: String, avroUpsertBucketName: String, batchSize: Int)(implicit cs: ContextShift[IO]) = {
+    system.actorOf(
+      AvroUpsertMonitorSupervisor.props(
+        FiniteDuration.apply(10, TimeUnit.MINUTES),
+        FiniteDuration.apply(10, TimeUnit.SECONDS),
+        workspaceService,
+        googleServicesDAO,
+        samDAO,
+        googleStorage,
+        googlePubSubDAO,
+        avroUpsertPubSubProject,
+        avroUpsertPubSubTopic,
+        avroUpsertPubSubSubscription,
+        avroUpsertBucketName,
+        batchSize,
+        1
```

Comment:
Should we be starting more than 1 of these workers? Having only 1 protects us from a 2nd worker picking up a message that another has been working on past the ack deadline, but also allows us to do only 1 import at a time.

Reply:
I am slightly concerned about having more than 1. I think the route we're going to go is to ack as soon as we load the JSON into memory, but PubSub isn't known for being an exactly-once messaging system so I'd be worried about duplicating heavy work with 2 workers and maybe running out of heap space at the same time.

As a compromise, I will make it configurable so that will buy us more time to test it out with 2 workers and then just change the config if we find that it's okay.

Reply:
I agree with your concern. I have been worried about the at-least-once delivery as well. Being able to handle only one of these imports at a time is an unfortunate consequence that I did not foresee. Given no other way to protect against that, I'm fine sticking with 1 worker for now. Low throughout is better than jeopardizing Rawls.

As an improvement, we could put a heartbeat signal in the job bucket to detect both repeated delivery and stalled jobs that need to be retried. Then we can optimize memory usage to make room for concurrent workers.

Reply:
I totally agree this is a problem - I'd like to solve it "for real" in follow-on work that would incorporate a real persistence store beyond writing files to a jobid subdirectoy. If we had a real database we could do things like check job id status easily for end users, list the user's jobs ... and have a place to persist internal semaphores like "the message for this job has already been seen"